### PR TITLE
Wenjie Fang - Api-muncher - Octos

### PR DESCRIPTION
**Pull request by wf1101**
# API Muncher
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
:------------- | :-------------
How did you go about exploring the Edamam API, how did you try querying the API? |  I first googled Edamam API and apply for developer id and key. I then read Edamam search API documentation and tried in Postman. 
Describe your API Wrapper.  How did you decide on the methods you created? |  From what is required by the project. It's supposed to return a list of recipes and display an individual recipe. That means two wrapper behaviors are needed: list_recipes and find_recipe.
Describe an edge case or failure case test you wrote for your API Wrapper. | Find a recipe DNE. This will raise an EdemamError. In the controller#show, the error is rescued and to display the error message and redirect back whatever page it was. 
Explain how VCR aids in testing an API. |  The only difference I feel is that I do not need to create yml or other data before call the method or path. 
What is the Heroku URL of your deployed application? |  https://tramuncher.herokuapp.com/
Provide a link to the Trello board you used |  https://trello.com/b/72Kb3OUs/api-muncher

### CODE REVIEW

File: app/views/layouts/application.html.erb [hunk 1]
```diff
@@ -0,0 +1,21 @@
+<!DOCTYPE html>
+<html lang="en">
+  <head>
+    <meta charset="utf-8" />
+    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
+
+    <title><%= content_for?(:title) ? yield(:title) : "Untitled" %></title>
```
Comment (CheezItMan) [inline app/views/layouts/application.html.erb:7]: Providing a page title would be good.

File: app/views/recipes/show.html.erb [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+  <h2>
+     <%= link_to @recipe.name, "#{@recipe.url}", target: :_blank %>
+  </h2>
+
+  <section class="recipe-info">
+    <%= image_tag(@recipe.image, alt: "food-image", class: "food-pic") %>
+    <section class="list-info">
+      <ul>
+        <h3 class="blue-mc">
+          Ingridients:
```
Comment (CheezItMan) [inline app/views/recipes/show.html.erb]: ingredients, careful with spelling

File: test/lib/edemam_api_wrapper_test.rb [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+        recipe = EdemamApiWrapper.find_recipe(uri)
+      end
+    end
+
+    it "raise an error when you try to find a recipe DNE" do
+      VCR.use_cassette("recipes") do
+        uri = "edb6djdka343dad33a25bfb2aa0613ccb2c2f545b30a5"
+        proc {
+          EdemamApiWrapper.find_recipe(uri)
+        }.must_raise ArgumentError
```
Comment (CheezItMan) [inline test/lib/edemam_api_wrapper_test.rb]: This should be
```ruby
 }.must_raise EdemamApiWrapper::EdemamError
```

File: test/controllers/recipes_controller_test.rb [hunk 4]
```diff
@@ -0,0 +1,40 @@
+require "test_helper"
+
+describe RecipesController do
```
Comment (CheezItMan) [inline test/controllers/recipes_controller_test.rb:3]: You also need a test for the homepage

**Comment by CheezItMan** [comment]
# API Muncher
## What We're Looking For

Feature | Feedback
:------------- | :-------------
**Core Requirements** | 
Git hygiene | Good number of commits and good commit messages
Comprehension questions | Check, VCR also reduces API calls and lets you even run the tests offline
**General** | 
Rails fundamentals (RESTful routing, use of named paths) | Your recipe `show` action is not RESTful.  It should be something like `/recipes/:id`.  You also have an extra route for the homepage.  
Semantic HTML | Check
Errors are reported to the user | You have error messages for invalid recipe URIs, no message if the search brings in 0 results.  
API Wrapper to handle the API requests | Check
Controller testing | Check
Lib testing | Check, with one failing test, see my notes.
Search Functionality | Check
List Functionality | Check
Show individual item functionality (link to original recipe opens in new tab) | The original recipe doesn't open in a new tab, but that's a minor issue.
**Styling** | 
Responsive layout | Check, good use of flexbox
List View shows 10 items at a time/pagination | Check
The app is styled to create an attractive user interface | Very nicely styled
**API Features** | 
The App attributes Edaman | Check
The VCR casettes do not contain the API key | Check
**External Resources** | 
Link to Trello Board | Check, but I don't have acces
Link to deployed app on Heroku | Check
**Overall** | Very nicely done.  You had some small errors in routing and testing, but overall you did a nice job.  Take a look at my inline comments and let me know any questions.

File: app/views/recipes/show.html.erb (hunk 2 above)
Comment (wf1101) [reply app/views/recipes/show.html.erb]: OOOOOOps! Thank you Chris!

Comment (wf1101) [reply app/views/recipes/show.html.erb]: I have ingredients spelled wrong everywhere in my project, that's why it did not break.

File: test/lib/edemam_api_wrapper_test.rb (hunk 3 above)
Comment (wf1101) [reply test/lib/edemam_api_wrapper_test.rb]: Thanks. I modified the EdemamError after I wrote the test...

Comment (wf1101) [reply test/lib/edemam_api_wrapper_test.rb]: Corrected  it.

File: app/views/recipes/show.html.erb (hunk 2 above)
Comment (wf1101) [reply app/views/recipes/show.html.erb]: Corrected it.

File: app/views/layouts/application.html.erb (hunk 1 above)
Comment (wf1101) [reply app/views/layouts/application.html.erb:7]: Thanks for your advice. Added title for each page.
